### PR TITLE
Analytical realspace multipole electrostatics

### PR DESCRIPTION
**Pull request by steffen-wedig**
This PR 1) adds closed-form real-space electrostatics for the Gaussian multipole densities ($l \le 1$) as `RealSpaceAnalyticalEnergy` and `RealSpaceAnalyticalElectrostaticFeatures`, fixing the rotational-equivariance violation of the displaced-charge finite-difference scheme (ACEsuit/mace#1719), and 2) vectorizes the batched pair-list builder. The finite-difference modules are left untouched and remain the ones constructed by `GTOElectrostaticEnergy` / `GTOElectrostaticFeatures` (see compatibility discussion at the end. )

## Why the finite-difference scheme breaks equivariance

The displaced-charge construction represents each $l=1$ Gaussian as monopoles displaced by a fixed offset $a$ along the **lab-frame** axes. This is a forward finite difference of the derivative identity below, so it carries an $O(a)$ error whose direction is tied to the lab frame: rotating a structure does not rotate the displacement directions, and the energy is not invariant. (Shrinking $a$ is no way out: the displaced charges scale as $\mu/a$, so small offsets lose the signal to catastrophic cancellation.)

## Mathematical background

**The two-centre Coulomb integral.** The real-space electrostatic energy of the model density is a sum over pairs of the Coulomb integrals between basis functions on the two atoms,

$$T_{lm,\,l'm'}(\mathbf R) \;=\; \iint \frac{\phi_{lm}(\mathbf r - \mathbf R_i)\,\phi_{l'm'}(\mathbf r' - \mathbf R_j)}{|\mathbf r - \mathbf r'|}\, d\mathbf r\, d\mathbf r' \;=\; \int \phi_{lm}(\mathbf r - \mathbf R_i)\, V_{l'm'}(\mathbf r - \mathbf R_j)\, d\mathbf r,$$

i.e. the electrostatic potential $V_{l'm'}$ of one basis function projected onto the other. The basis functions are the solid-harmonic Gaussians $\phi_{lm}(\mathbf r) = C_{l\sigma}\, e^{-r^2/2\sigma^2}\, r^l\, Y_{lm}(\hat{\mathbf r})$ of `docs/README.md`. For the $s$–$s$ case ($l = l' = 0$) the potential of a unit Gaussian charge is $V_\sigma(r) = \text{erf}(r/\sqrt2\sigma)/r$, and projecting it onto a second Gaussian is easiest in Fourier space: with the Coulomb kernel $4\pi/k^2$ and Gaussian form factors $e^{-k^2\sigma^2/2}$, the two form factors multiply into a single Gaussian of variance $\Sigma^2 = \sigma_1^2 + \sigma_2^2$, so

$$B_0(R) \;=\; \int \frac{d^3k}{(2\pi)^3}\, \frac{4\pi}{k^2}\, e^{-k^2\Sigma^2/2}\, e^{i\mathbf k\cdot\mathbf R} \;=\; \frac{\text{erf}(R/\sqrt{2}\,\Sigma)}{R}.$$

This combined-width rule is why one kernel family serves both the energy (equal widths, $\Sigma = \sqrt2\,\sigma_{\text{charge}}$) and the feature projection (mixed widths, $\Sigma_n^2 = \sigma_{\text{charge}}^2 + \sigma_{\text{field},n}^2$ per radial channel).

**Hobson's theorem.** The regular solid harmonics $\mathcal R_{lm}(\mathbf r) = r^l Y_{lm}(\hat{\mathbf r})$ satisfy $\mathcal R_{lm}(\nabla)\, e^{-\alpha r^2} = (-2\alpha)^l\, \mathcal R_{lm}(\mathbf r)\, e^{-\alpha r^2}$: every solid-harmonic Gaussian is a solid-harmonic *derivative* of the $s$-Gaussian of the same width. For $l = 1$ this is elementary, $r_\alpha\, e^{-r^2/2\sigma^2} = -\sigma^2\, \partial_\alpha\, e^{-r^2/2\sigma^2}$. Substituting this on both centres of the Coulomb integral and integrating by parts moves all derivatives onto the centre separation, giving the master formula (in moment normalization, where a monopole coefficient is a charge in $|e|$ and a dipole coefficient a moment in $|e|\cdot$Å)

$$T_{l_1 m_1,\, l_2 m_2}(\mathbf R) \;=\; (-1)^{l_1}\, \mathcal R_{l_1 m_1}(\nabla_{\mathbf R})\, \mathcal R_{l_2 m_2}(\nabla_{\mathbf R})\; B_0(R).$$

The displaced-charge scheme is exactly the first-order finite difference of these derivatives; the analytical modules evaluate them in closed form.

**Explicit $l \le 1$ tensors and the Boys functions.** Differentiating a radial kernel generates the chain $B_{n+1}(R) = -\tfrac{1}{R}\tfrac{dB_n}{dR}$, so all $l \le 1$ tensors are built from $B_0, B_1, B_2$ with the angular structure carried only by $\mathbf R$ and $\delta_{\alpha\beta}$. Writing $t = R^2/2\Sigma^2$, these kernels are Boys functions in disguise:

$$B_n(R) = 2\sqrt{\tfrac{a}{\pi}}\,(2a)^n\, F_n(t), \qquad F_n(t) = \int_0^1 u^{2n} e^{-t u^2}\, du, \qquad a = \tfrac{1}{2\Sigma^2},\; t = aR^2$$

 the same objects that molecular-integral libraries (e.g. libcint's `int2c2e`) evaluate; our case is the simple two-centre special case. The resulting pair energy for moment-normalized coefficients is

$$E_{ij} = q_i q_j B_0 - (q_i\boldsymbol\mu_j - q_j\boldsymbol\mu_i)\cdot\mathbf{R}\, B_1 + (\boldsymbol\mu_i\cdot\boldsymbol\mu_j) B_1 - (\boldsymbol\mu_i\cdot\mathbf{R})(\boldsymbol\mu_j\cdot\mathbf{R})\, B_2,$$

with $\mathbf R = \mathbf R_j - \mathbf R_i$: the last two terms are the exact dipole–dipole interaction of two Gaussian dipole densities — the familiar point tensor with radial factors $1/R^3,\ 3/R^5$ replaced by $B_1, B_2$. The formula is exact at **all** separations (overlapping densities included, no multipole expansion is made) and exactly rotationally equivariant by construction. For $R \gg \Sigma$ it converges exponentially to the classical point-multipole tensors, and at $R \to 0$ it reproduces the analytic self-interaction constants ($B_n(0) = \sqrt{2/\pi}\,\Sigma^{-(2n+1)}/(2n+1)$), which match the independently integrated `GTOSelfInteractionBlock` to ~1e-11.

The projected-potential features use the same kernels with the mixed widths $\Sigma_n$: the scalar feature is the smeared potential $q_j B_0 + (\boldsymbol\mu_j\cdot\mathbf D) B_1$ and the $l=1$ feature is its exact gradient with respect to the receiver position — no probe displacement, and no offset-dependent normalization factors.

**Numerics: why the evaluation branches between a recursion and a series.** In closed form, $B_0 = \text{erf}(R/\sqrt2\Sigma)/R$ and the higher kernels follow the upward recursion $B_1 = (B_0 - G)/R^2$, $B_2 = (3B_1 - G/\Sigma^2)/R^2$ with the Gaussian $G(R) = \sqrt{2/\pi}\,\Sigma^{-1} e^{-t}$. For $t \ll 1$ both operands of the subtraction approach the same constant $\sqrt{2/\pi}/\Sigma$ and their difference is only $O(t)$ times their magnitude, so with machine precision $\varepsilon$ the relative error of $B_1$ is $\sim \varepsilon/t$ and — after the second subtraction — that of $B_2$ is $\sim \varepsilon/t^2$ (catastrophic cancellation; at $R = 0$ the recursion is literally $0/0$ although $B_n(0)$ is finite). This is not an exotic corner: with $\sigma = 1.5$ Å every bonded pair sits at $t \lesssim 0.1$, where float32 ($\varepsilon \approx 6\cdot10^{-8}$) retains only $\sim\!10^{-5}$ relative accuracy in $B_2$ and degrades to pure noise below $t \approx 10^{-3}$. Below $t^\ast = 0.5$ each $B_n$ is therefore evaluated from its own Boys-function Taylor series,

$$B_n(R) = \sqrt{\tfrac{2}{\pi}}\; \Sigma^{-(2n+1)} \sum_{k} \frac{(-t)^k}{k!\,(2n+2k+1)},$$

which contains no subtraction of near-equal quantities (the $k=0$ term dominates), is accurate to $\sim\varepsilon$ independent of $t$, and reduces to the exact constants at $R = 0$. The series cannot be used everywhere either: it is an alternating sum whose terms grow like $t^k/k!$ before decaying, so for $t \gg 1$ it develops its own cancellation while the closed form loses at most a fraction of a digit for $t \ge 0.5$ — the two formulas have complementary stability regions. With 13 terms the series truncation at the crossover is $< 0.5^{13}/13! \approx 2\cdot10^{-14}$, and the branches agree to $5\cdot10^{-12}$ in values *and* derivatives across the overlap window. The dispatch is a clamped `torch.where` (autograd evaluates both branches, so each branch's argument is clamped into its safe domain first): values and first/second derivatives are finite and correct everywhere, training with force losses (double backward) is NaN-free by construction, and float32 relative accuracy is $\lesssim 10^{-6}$ across the full distance range.

## Correctness tests

`tests/test_realspace_analytical_electrostatics.py` validates the implementation in four independent layers:

- **Kernel oracles**: values against symbolic (sympy) derivatives over $t \in [10^{-6}, 10^2]$; series/closed-form branch consistency in values and derivatives; autograd first and second derivatives of the dispatched kernels across and exactly at the crossover (via $dB_n/ds = -B_{n+1}/2$, $s = R^2$); zero-distance and far-field limits; float32 accuracy.
- **First-principles quadrature**: the full pair formula against 3D Gauss–Hermite quadrature (unequal widths, overlapping to separated), and the feature module against a numerical projection of the source potential onto receiver basis functions built directly from `get_Cl_sigma` (both normalizations) — pinning the mixed-width kernels, normalization factors, sign convention, and e3nn $(y,z,x)$ ordering independently of the existing code.
- **Consistency with the existing modules**: exact agreement with the (already exact) $l=0$ path; first-order Richardson extrapolation of the finite-difference modules converges to the analytical result and the finite-difference error scales as $O(a)$; self terms match `GTOSelfInteractionBlock`.
- **Symmetry and gradients**: rotation equivariance (energy to $10^{-12}$, features transform correctly), translation/permutation invariance, batch semantics, `gradcheck`/`gradgradcheck` including systems with pair distances straddling both kernel branches.

## Vectorized pair-list builder

`batch_complete_graph_excluding_self_duplicates_vector` no longer loops over graphs in Python. The construction sorts duplicated nodes by graph once and then either builds a batched broadcast mesh (when all graphs have equal size: single systems and uniform batches) or enumerates each sender's receiver block with `repeat_interleave` and gathers — deliberately avoiding int64 division/modulo on the dense-edge tensor, which is emulated and slow on GPUs. Edge lists are bit-identical to the previous implementation.

B200 timings versus the per-graph loop ($N$ = duplicates per node; $N=4$ is the finite-difference regime, $N=1$ the analytical one):

| batch | graphs | atoms | N=1 | N=4 |
|---|---|---|---|---|
| 256 mixed water clusters | 256 | 4368 | **100x** | **89x** |
| training batch, 32 random clusters | 32 | 769 | 13x | 14x |
| training batch, 128 random clusters | 128 | 2750 | **52x** | **49x** |
| 127 clusters + one 3000-atom droplet | 128 | 5737 | 23x | 3.0x |
| 16 x 192-atom droplets | 16 | 3072 | 13x | 5.2x |
| single 3000-atom droplet | 1 | 3000 | 1.0x | 1.0x |

## Module benchmarks (B200, float32)

Finite-difference vs analytical, matching the MACE-POLAR-1 hyperparameters ($\sigma = 1.5$, feature widths $[1.5, 3.0]$, receiver normalization). The analytical path also shrinks the interaction graph ~16x (no displaced pseudo-particles).

| system | module | pass | finite difference | analytical | speedup |
|---|---|---|---|---|---|
| 3000-atom droplet | energy | forward+forces | 90.8 ms / 12.7 GB | 12.8 ms / 2.2 GB | **7.1x** |
| 3000-atom droplet | features | forward+forces | 98.5 ms / 13.3 GB | 21.5 ms / 4.2 GB | **4.6x** |
| 16 x 192-atom droplets | energy | forward+forces | 6.9 ms / 0.9 GB | 3.6 ms / 0.2 GB | 1.9x |
| 256 small clusters | energy | forward+forces | 3.0 ms / 0.2 GB | 3.7 ms / 0.06 GB | 0.8x |

For batches of very small graphs the analytical modules are marginally slower (launch-overhead bound); everywhere else they win, increasingly so with system size, and always with several-fold less memory.

## Impact on MACE-POLAR

**If merged, this affects MACE-POLAR**, which builds its real-space electrostatics from this package. Model-level measurements with the released MACE-POLAR-1 foundation weights (float64, rotations about the centroid, non-periodic systems — periodic ones take the reciprocal-space branch, which already uses exact analytic form factors):

- **Equivariance**: with the finite-difference modules, the M and L models violate rotational invariance by 1–8.5 meV in energy and 2–13 meV/Å in forces (largest on a 192-molecule SO2 droplet — strong molecular dipoles, the system class where the bug was found). Hot-swapping the analytical modules reduces the violation by 4–6 orders of magnitude, to the numerical floor of the rest of the network (~$10^{-12}$ relative). The S model doesnt have the l=1 dipoles, so its unaffected.
- **Prediction shifts**: swapping the analytical modules into the *trained* weights changes predictions — up to ~0.26 meV/atom in energy on charged water clusters and up to ~38 meV/Å in forces on a 3000-atom SO2 droplet for the M model (S: only ~0.0004 meV/atom).

## Compatibility recommendation: do not swap for existing models

The released foundation models were **trained with** the finite-difference electrostatics, so their weights implicitly compensate its $O(a)$ errors; swapping the analytical modules under them breaks those learned corrections (the prediction shifts above). We therefore deliberately do **not** change what `GTOElectrostaticEnergy` / `GTOElectrostaticFeatures` construct:

- **existing checkpoints** keep the finite-difference modules and reproduce their training behavior exactly;
- **newly trained models** should use the analytical modules, exactly equivariant, exact at all separations, faster and lighter at scale. 
For now, the Analytical modules is not tied in anywhere though. 

A natural follow-up is a constructor flag on the two wrapper classes (`finite_difference` | `analytical`, defaulting to `finite_difference` for checkpoint compatibility), so new training runs can opt in with one argument. Can add this here or just leave it for new training runs.

### CODE REVIEW

**Comment by steffen-wedig** [comment]
- Merged the latest main into this branch
- ensured that the external field electrostatics raise on the analytical realspace electorstatics, because there are some changes required there. 
- Added more comments explaining the work and todos for more speedups.
- I also added a argument to GTOElectrostaticEnergy/Features to allow toggling between analytical and finite_difference realspace behaviour. 

Models can be converted by 
```
import torch
torch.serialization.add_safe_globals([slice])

model = torch.load("MACE-POLAR-1-S.model", map_location="cpu", weights_only=False)
model.coulomb_energy.set_realspace_method("analytical")
model.electric_potential_descriptor.set_realspace_method("analytical")
torch.save(model, "MACE-POLAR-1-S-analytical.model")
```

but this changes the models predictions, so it shouldnt be done lightly (i.e. without additional finetuning). 

Sorry for the heavy PR, 75% of it are tests though : /

**Comment by steffen-wedig** [comment]
Merging this unblocks the mace nvalchemitoolkit ops electrostatics backend [PR](https://github.com/ACEsuit/mace/pull/1524).  Energies in the AnalyticalEnergy and nvalchemi toolkitops agree to 1e-4 meV/atom , and the residual seems to sit in their [approximation of erfc](https://github.com/NVIDIA/nvalchemi-toolkit-ops/blob/32445c5f00e83fccd475589f1a6f5ac9e81e6b02/nvalchemiops/math/math.py#L98)
